src/libraries/System.Private.CoreLib/src/System/IO/FileSystem.TryCloneFile.OSX.cs

```diff
@@ -0,0 +1,87 @@
+using System.Diagnostics;
+
+namespace System.IO
+{
+    internal static partial class FileSystem
+    {
+        private static bool TryCloneFile(string sourceFullPath, string destFullPath, bool overwrite)
+        {
+            // Try to clone the file immediately, this will only succeed if the
+            // destination doesn't exist, so we don't worry about locking for this one.
+            int flags = Interop.@libc.CLONE_ACL;
+            retry1:
+            if (Interop.@libc.clonefile(sourceFullPath, destFullPath, flags) == 0)
+            {
+                // Success.
+                return true;
+            }
+            Interop.Error error = Interop.Sys.GetLastError();
+
+            // Some filesystems don't support ACLs, so may fail due to trying to copy ACLs.
+            // This will disable them and allow trying again (a maximum of 1 time).
+            if (flags != 0 && error == Interop.Error.EINVAL)
+            {
+                flags = 0;
+                goto retry1;
+            }
+
+            // Try to delete the destination file if we're overwriting.
+            if (error == Interop.Error.EEXIST && overwrite)
+            {
+                // Delete the destination. This should fail on directories. Get a lock to the dest file to ensure we don't copy onto it when
+                // it's locked by something else, and then delete it. It should also fail if destination == source since it's already locked.
+                try
+                {
+                    using SafeFileHandle? dstHandle = SafeFileHandle.Open(destFullPath, FileMode.Open, FileAccess.ReadWrite,
+                        FileShare.None, FileOptions.None, preallocationSize: 0);
+                    if (Interop.Sys.Unlink(destFullPath) < 0)
+                    {
+                        Interop.Error errorInfo = Interop.Sys.GetLastError();
+                        if (error != Interop.Error.ENOENT)
+                        {
+                            // Fall back to standard copy as an unexpected error has occurred.
+                            return false;
+                        }
+                    }
+                }
+                catch (FileNotFoundException)
+                {
+                    // We don't want to throw if it's just the file not existing, since we're trying to delete it.
+                }
+
+                // Try clonefile now we've deleted the destination file.
+                retry2:
+                if (Interop.@libc.clonefile(sourceFullPath, destFullPath, flags) == 0)
+                {
+                    // Success.
+                    return true;
+                }
+                error = Interop.Sys.GetLastError();
+
+                // Some filesystems don't support ACLs, so may fail due to trying to copy ACLs.
+                // This will disable them and allow trying again (a maximum of 1 time).
+                if (flags != 0 && error == Interop.Error.EINVAL)
```

Comment:
```suggestion
                if (error == Interop.Error.EINVAL && flags != 0)
```
nit, easier to read imo